### PR TITLE
Add support for TEXT SEARCH CONFIGURATION objects

### PR DESCRIPTION
**Pull request by thanodnl**
DESCRIPTION: Implement TEXT SEARCH CONFIGURATION propagation

The change adds support to Citus for propagating TEXT SEARCH CONFIGURATION objects. TSConfig objects cannot always be created in one create statement, and instead require a create statement followed by many alter statements to get turned into the object they should represent.

To support this we add functionality to the worker to create or replace objects based on a list of statements. When the lists of the local object and the remote object correspond 1:1 we skip the creation of the object and simply mark it distributed. This is especially important for TSConfig objects as initdb pre-populates databases with a dozen configurations (for many different languages).

When the user creates a new TSConfig based on the copy of an existing configuration there is no direct link to the object copied from. Since there is no link we can't simply rely on propagating the dependencies to the worker and send a qualified

TODO
 - [x] add owner change to idempotent creation list
 - [x] rebase after #5704 lands to dedupe ensure sequential code directly
 - [x] add a test that verifies we don't recreate initdb configurations when they have _not_ been altered #5703

### CODE REVIEW

File: src/backend/distributed/commands/text_search.c [hunk 1]
```diff
@@ -0,0 +1,804 @@
...
+	}
+
+	/* we haven't found the token */
+	ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("token type \"%d\" does not exist in parser", tokentype)));
+}
+
+
+static List *
+GetTSParserNameList(Oid tsparserOid)
```
Comment (thanodnl) [inline src/backend/distributed/commands/text_search.c:588]: revisit name of function
```suggestion
get_ts_parser_namelist(Oid tsparserOid)
```

File: src/backend/distributed/commands/text_search.c [hunk 2]
```diff
@@ -0,0 +1,804 @@
...
+ *
+ * text_search.c
+ *    Commands for creating and altering TEXT SEARCG objects
+ *
+ * Copyright (c) Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
```
Comment (thanodnl) [inline src/backend/distributed/commands/text_search.c:11]: many functions in file without comment, please add

File: src/backend/distributed/deparser/qualify_text_search_stmts.c [hunk 3]
```diff
@@ -0,0 +1,225 @@
+#include "postgres.h"
```
Comment (thanodnl) [inline src/backend/distributed/deparser/qualify_text_search_stmts.c:1]: missing copyright

File: src/backend/distributed/worker/worker_create_or_replace.c [hunk 4]
```diff
@@ -68,6 +69,8 @@ WrapCreateOrReplace(const char *sql)
  *  - non destructive if there is data store that would be destroyed if the object was
  *    used in a table on this node, eg. types. If the type would be dropped with a cascade
  *    it would drop any column holding user data for this type.
+ *
+ *  TODO: we can probably remove the existing object if it has no dependents
```
Comment (thanodnl) [inline src/backend/distributed/worker/worker_create_or_replace.c:73]: This deserves a bit of a deeper discussion.

I feel like we need to refactor `worker_create_or_replace_object` in a way that it can optionally take a list of statements and reconciles the object if possible.

`TEXT SEARCH CONFIGRURATION` can be created as follows:

```sql
CREATE TEXT SEARCH CONFIGURATION foo ( copy = dutch );
```

From the catalog the relation to the `pg_catalog.dutch` text search configuration is lost. Instead it used said configuration to copy its parser and all its mappings. Hence we will create the object on the workers like the following:

```
CREATE TEXT SEARCH CONFIGURATION public.foo ( parser = pg_catalog.default );
ALTER TEXT SEARCH CONFIGURATION public.foo ADD MAPPING FOR token1 WITH dict1, dict2;
ALTER TEXT SEARCH CONFIGURATION public.foo ADD MAPPING FOR token2 WITH dict3, dict;
...
```

All statements combined form the state of the object.
Ideally we would verify all statements are the same between coordinator and worker and do nothing (except mark them distributed eventually). This _should_ work well for all configurations installed with initdb as well.

If settings differ we could still fallback on old behaviour to rename the old object and recreate according to spec.
And maybe drop the object if it wasn't used anyway, like the TODO describes.

The dropping is trivial, but not related to this PR, so might perform this in a separate PR.
The refactor of worker replace is actually related to this PR, but sizeable work, although I think a requirement to _not_ pollute catalogs on the workers when initdb created configs are used in tables/indices.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 5]
```diff
@@ -102,7 +102,6 @@ static GrantStmt * GenerateGrantStmtForRights(ObjectType objectType,
 											  bool withGrantOption);
 static List * GetObjectsForGrantStmt(ObjectType objectType, Oid objectId);
 static AccessPriv * GetAccessPrivObjectForGrantStmt(char *permission);
-static RoleSpec * GetRoleSpecObjectForGrantStmt(Oid roleOid);
```
Comment (thanodnl) [inline src/backend/distributed/metadata/metadata_sync.c:105]: funtimes, enterprise has these definitions reordered, hence the merge conflict :/
couldn't find a way to change this without conflict. Only by keeping this function and using a copy for my purpose, but that seems wrong. Will fix conflicts after approval.

File: src/backend/distributed/worker/worker_create_or_replace.c (hunk 4 above)
Comment (thanodnl) [reply src/backend/distributed/worker/worker_create_or_replace.c:73]: a potential way to prevent initdb created configuration to be overwritten with semantically the same object: https://github.com/citusdata/citus/pull/5703

File: src/test/regress/sql/text_search.sql [hunk 6]
```diff
@@ -0,0 +1,208 @@
+CREATE SCHEMA text_search;
+CREATE SCHEMA text_search2;
+SET search_path TO text_search;
+
+-- create a new configruation from scratch
+CREATE TEXT SEARCH CONFIGURATION my_text_search_config ( parser = default );
```
Comment (onderkalaci) [inline src/test/regress/sql/text_search.sql:6]: Given that creation logic of Citus local tables are a little different than dist / ref table, it would be useful to create one citus local table and check if the shard also has the index:
```SQL

...
-- require coordinator in the metadata
SELECT citus_add_local_table_to_metadata('t1');


\d t1
              Table "text_search.t1"
 Column |  Type   | Collation | Nullable | Default 
--------+---------+-----------+----------+---------
 id     | integer |           |          | 
 name   | text    |           |          | 
Indexes:
    "t1_search_name" gin (to_tsvector('my_text_search_config'::regconfig, COALESCE(name, ''::character varying::text)))
\d t1_102009
           Table "text_search.t1_102009"
 Column |  Type   | Collation | Nullable | Default 
--------+---------+-----------+----------+---------
 id     | integer |           |          | 
 name   | text    |           |          | 
Indexes:
    "t1_search_name_102009" gin (to_tsvector('my_text_search_config'::regconfig, COALESCE(name, ''::character varying::text)))

```

We could consider `undistribute_table` as well as it also has some edge cases

File: src/test/regress/expected/text_search.out [hunk 7]
```diff
@@ -0,0 +1,390 @@
+CREATE SCHEMA text_search;
+CREATE SCHEMA text_search2;
+SET search_path TO text_search;
+-- create a new configruation from scratch
+CREATE TEXT SEARCH CONFIGURATION my_text_search_config ( parser = default );
+CREATE TABLE t1(id int, name text);
+CREATE INDEX t1_search_name ON t1 USING gin (to_tsvector('text_search.my_text_search_config'::regconfig, (COALESCE(name, ''::character varying))::text));
```
Comment (onderkalaci) [inline src/test/regress/expected/text_search.out:7]: We might consider adding a simple test to arbitrary configs, it sometimes spots very interesting edge cases

File: src/backend/distributed/commands/index.c [hunk 8]
```diff
@@ -747,14 +741,25 @@ PostprocessIndexStmt(Node *node, const char *queryString)
...
 
+	Oid indexRelationId = get_relname_relid(indexStmt->idxname, schemaId);
+
+	/* ensure dependencies of index exist on all nodes */
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, RelationRelationId, indexRelationId);
+	EnsureDependenciesExistOnAllNodes(&address);
+
+	/* furtheron we are only processing CONCURRENT index statements */
+	if (!indexStmt->concurrent)
```
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/commands/index.c:752]: I was hoping this had been tested before. The logic for this was already in place, just refactored slightly.

Can you describe  in more detail what new tests would add to this?

File: src/backend/distributed/commands/text_search.c [hunk 9]
```diff
@@ -0,0 +1,984 @@
...
+ * there is no comment on a configuration.
+ *
+ * The reason for a list is for easy use when building a list of all statements to invoke
+ * to recreate the text search configuration. An empty list can easily be concatinated
+ * without inspection, contrary to a NULL ptr if we would return the CommentStmt struct.
+ */
+static List *
+GetTextSearchConfigCommentStmt(Oid tsconfigOid)
+{
+	char *comment = GetComment(tsconfigOid, TSConfigRelationId, 0);
```
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/commands/text_search.c:484]: Makes a lot of sense, I implemented it here because the comments are quite prominently shown on `\dF` and pre-populated by initdb.

File: src/backend/distributed/commands/text_search.c [hunk 10]
```diff
@@ -0,0 +1,984 @@
...
+
+/*
+ * GetTextSearchConfigMappingStmt returns a list of all mappings from token_types to
+ * dictionaries configured on a text search configuration identified by tsconfigOid.
+ *
+ * Many mappings can exist on a configuration which all require their own statement to
+ * recreate.
+ */
+static List *
+GetTextSearchConfigMappingStmt(Oid tsconfigOid)
```
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/commands/text_search.c:507]: I have also observed that postgres' errors around ts configs are quite cryptic and can reproduce the above without object propagation. Although for me there is an entry for which the index conflicted. Wonder how it happened for you that they weren't visible.

File: src/backend/distributed/commands/text_search.c [hunk 11]
```diff
@@ -0,0 +1,984 @@
...
+	 * stmts list.
+	 */
+	if (stmt != NULL)
+	{
+		stmts = lappend(stmts, stmt);
+		stmt = NULL;
+	}
+
+	systable_endscan_ordered(mapscan);
+	index_close(mapidx, AccessShareLock);
```
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/commands/text_search.c:577]: I copied this from posgres' [`lookup_ts_config_cache`](https://github.com/postgres/postgres/blob/50e5bc582a4371a88218078a2c651d901bf87d96/src/backend/utils/cache/ts_cache.c#L519-L521) function.
I will change it now though, as I do remember we keep them for most of the objects.
I wonder though if that is required here or not.

File: src/backend/distributed/commands/text_search.c [hunk 12]
```diff
@@ -0,0 +1,984 @@
...
+	return address;
+}
+
+
+/*
+ * AlterTextSearchConfigurationSchemaStmtObjectAddress resolves the ObjectAddress for the
+ * TEXT SEARCH CONFIGURATION being moved to a different schema. Optionally errors if the
+ * configuration does not exist based on the missing_ok flag passed in by the caller.
+ *
+ * This can be called, either before or after the move of schema has been executed, hence
```
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/commands/text_search.c:802]: Correct

File: src/backend/distributed/commands/text_search.c [hunk 13]
```diff
@@ -0,0 +1,984 @@
...
+
+	char *newName = palloc0(NAMEDATALEN);
+	char suffix[NAMEDATALEN] = { 0 };
+	char *baseName = rel->relname;
+	int baseLength = strlen(baseName);
+	int count = 0;
+
+	while (true)
+	{
+		int suffixLength = SafeSnprintf(suffix, NAMEDATALEN - 1, "(citus_backup_%d)",
```
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/commands/text_search.c:909]: Probably not, it is pretty hard to achieve :D Will add a test.

However, the thing we talked about earlier, to actually drop existing objects if they are not used will make this code even less likely to start hitting, but we will see.

File: src/test/regress/sql/text_search.sql (hunk 6 above)
Comment (thanodnl) [reply src/test/regress/sql/text_search.sql:6]: I wonder if this should be part of this PR.
This PR doesn't implement the creation of the indexes, that was already done correct.
Instead this PR purely adds the propagation of the object to remote machines.

Not saying this shouldn't be tested, but I guess this should be part of a more holistic test suite that tests all possible index creation syntax

File: src/backend/distributed/commands/index.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/commands/index.c:752]: Thinking a bit about this, I will add 1 test, just to be sure.

One that will have the TSConfig object propagate for creation of indexes concurrently. I expect them to work, but might be good to verify

File: src/backend/distributed/commands/text_search.c (hunk 10 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/text_search.c:509]: I might report these to pg-hackers

File: src/backend/distributed/commands/text_search.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/text_search.c:911]: On the worker:
```SQL
set citus.enable_metadata_Sync to off;

CREATE TEXT SEARCH CONFIGURATION my_text_search_config_test ( copy = dutch );
```
On the coordinator:
```SQL
CREATE TEXT SEARCH CONFIGURATION my_text_search_config_test ( parser = default );
```

Now, on the worker I can see:
```SQL
 select cfgname from pg_ts_config WHERE cfgname ilike '%my_text_search_config_test%';
┌────────────────────────────────────────────┐
│                  cfgname                   │
├────────────────────────────────────────────┤
│ my_text_search_config_test(citus_backup_0) │
│ my_text_search_config_test                 │
└────────────────────────────────────────────┘
(2 rows)

```

[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/commands/text_search.c:911]: (non-sense test, just to trigger this code-path which is useful anyway)

File: src/backend/distributed/metadata/dependency.c [hunk 14]
```diff
@@ -1048,6 +1066,28 @@ GetRelationStatsSchemaDependencyList(Oid relationId)
...
+	*oids = lappend_oid(*oids, formPgIndex->indexrelid);
+}
+
+
+/*
+ * GetRelationIndicesDependencyList creates a list of ObjectAddressDependencies for the
+ * indexes on a given relation.
+ */
+static List *
+GetRelationIndicesDependencyList(Oid relationId)
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/dependency.c:1084]: as a nice side-effect of this, functions on the indexes are also supported now:
```SQL
CREATE OR REPLACE FUNCTION lower_sql(a text)                                   RETURNS text                                                                                                         LANGUAGE sql AS $$ select lower($1); $$;
create table test(a text);
select create_distributed_table('test', 'a');
 create index i1 on test (lower_sql(a));


SELECT pg_identify_object_as_address(classid, objid, objsubid) from citus.pg_dist_object;
...
│ (function,"{public,lower_sql}",{pg_catalog.text})                      │
...

```

Whereas on the master branch this fails:
```SQL
RROR:  function lower_sql(text) does not exist
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
CONTEXT:  while executing command on localhost:9700
```


We are already making `CREATE FUNCTION` to propagate soon via #5701, but still it is good for other objects if can be used in indexes

**Review by onderkalaci — APPROVED** [review]
I don't really have any further comments other than those I shared.

Given that name quotations are tricky, I think using a quoted schema name in the tests is a good one that we should consider changing the it in the test file you added. 

Other reviews are optional, you may decide to follow or not.

File: src/backend/distributed/commands/index.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/commands/index.c:752]: test has been added. Found snapshot leaking due to this comment, thank you!

File: src/backend/distributed/commands/text_search.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (thanodnl) [reply src/backend/distributed/commands/text_search.c:911]: I think it's a good one to add. I will actually use the same to verify the behaviour of _not_ renaming the object if the configurations are the same.

However, a future change in not _renaming_ but removing when an object like that is not in use will break this test again :D

File: src/test/regress/expected/text_search.out (hunk 7 above)
Comment (thanodnl) [reply src/test/regress/expected/text_search.out:7]: There is a thread lingering for me to add tests to this suite for many of the object propagation things. Am planning to address those after codefreeze.

File: src/backend/distributed/metadata/dependency.c (hunk 14 above)
Comment (hanefi) [reply src/backend/distributed/metadata/dependency.c:1084]: I think that it is worth adding to changelog for Citus 11.

How about `Adds support for functions on index definitions` ?
